src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.PerAnalyzerState.cs

```diff
@@ -199,15 +199,29 @@ private void MarkSyntaxTreeProcessed_NoLock(SyntaxTree tree)
                 _lazySyntaxTreesWithAnalysisData[tree] = AnalyzerStateData.FullyProcessedInstance;
             }
 
+            private Dictionary<int, DeclarationAnalyzerStateData> EnsureDeclarationDataMap_NoLock(ISymbol symbol, Dictionary<int, DeclarationAnalyzerStateData> declarationDataMap)
+            {
+                Debug.Assert(_pendingDeclarations[symbol] == declarationDataMap);
+
+                if (declarationDataMap == null)
+                {
+                    declarationDataMap = _currentlyAnalyzingDeclarationsMapPool.Allocate();
```

Comment:
What's the threading story here? Can multiple threads call `TryStartAnalyzingDeclaration_NoLock` at the same time?

Reply:
No, methods suffixed with `_NoLock` are all inside a lock in this file.

Reply:
:bulb: You can include a debug assertion that a monitor is held at the beginning of the NoLock methods to aid with documentation.